common/leica_sep.c

```diff
@@ -66,23 +70,89 @@ static int leica_sep_transfer(struct leica_sep_funcs *f, char *cmd, char *data,
 	return bytes_received;
 }
 
-int leica_sep_get_board_id(struct leica_sep_funcs *f, struct leica_sep_board_id *board_id)
+/*
+ * leica_get_sep_payload
+ *
+ * Function that takes as argument the received data from the MCU and extracts
+ * the payload. The result is stored overwriting the received argument with a
+ * new string.
+ */
+static int leica_get_sep_payload(char *data, size_t size)
 {
-	char data[512];
-	int n_bytes;
-	size_t n_header = sizeof(SEP_BOARD_ID) - 1 + sizeof(",XXXX") - 1;
+	int idx = 0, new_idx = 0, quotes = 0, column = 0;
+	while(idx < size && data[idx] != '\r')
+	{
+		if (data[idx] == ':')
+		{
+			column = 1;
+		}
+		else if (column && data[idx] == '"')
+		{
+			quotes++;
+			if (quotes == 2) break;
+		}
+		else if(column)
+		{
+			data[new_idx] = data[idx];
+			new_idx++;
+		}
+		idx++;
+	}
+	if (!column) return -1;
+	/*We substitute the last char with '\0' to terminate the payload string*/
+	data[new_idx] = '\0';
+	return 0;
+}
 
-	if(!board_id)
-		return -1;
+/*
+ * leica_get_sep
+ *
+ * Function to communicate with the MCU and get parsed data from the SEP
+ * payload. The result is stored in an environment variable named <sep_name>.
+ * Available sep names are: 'board_id', 'production_flag'.
+ */
+int leica_get_sep(char *sep_name)
+{
+	char data[SEP_LENGTH];
+	int n_bytes, i = 0;
 
-	n_bytes = leica_sep_transfer(f, SEP_BOARD_ID, data, sizeof(data), 500);
-	if (n_bytes < (int)(n_header + sizeof(*board_id))) {
-		printf("[SEP DBG]: Receive failed. Using test-data... Remove this!!! \n");
-		memcpy(data, "SR111,abcd:\"916761A2107310001\"\r\n", sizeof("SR111,abcd:\"916761A2107310001\"\r\n") - 1);
-		//return -1;
+	while(i < num_seps)
+	{
+		if (!strcmp(sep_name, leica_seps[i].name))
+		{
+			n_bytes = leica_sep_transfer(&sep_funcs, (char *)leica_seps[i].sep, data, SEP_LENGTH, 500);
```

Comment:
Please use sizeof(data) - if we want to change the size we only need to do it in declaration of data

Reply:
Why is the typecast required?